dataikuapi/dss/ml.py

```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+            assert all(val <= limit_max for val in values), "Value(s) above hyperparameter \"{}\" limit {}".format(hyperparameter_name, limit_max)
+        if len(set(values)) < len(values):
+                warnings.warn("Detected duplicates in provided values: " + str(sorted(values)))
+        self[hyperparameter_name]["values"] = values
+
+        if mode_as_explicit:
+            self._set_hyperparameter_mode_as_explicit(hyperparameter_name)
+
+    def _set_numerical_range(self, hyperparameter_name, range_min=None, range_max=None, nb_values=None, mode_as_range=True):
+        self._check_hyperparameter_name(hyperparameter_name)
+
+        if range_min is None and range_max is None and nb_values is None:
+            warnings.warn("Numerical range for hyperparameter \"{}\" not modified".format(hyperparameter_name))
+        else:
+            if range_min is not None:
+                self._check_range_bound(hyperparameter_name, range_min)
+                limit_min = self[hyperparameter_name]["limit"].get("min")
+                if limit_min is not None:
+                    assert limit_min <= range_min, "Range min {} is below hyperparameter \"{}\" limit {}".format(range_min, hyperparameter_name, limit_min)
+            if range_max is not None:
+                self._check_range_bound(hyperparameter_name, range_max)
+                limit_max = self[hyperparameter_name]["limit"].get("max")
+                if limit_max is not None:
+                    assert range_max <= limit_max, "Range max {} is above hyperparameter \"{}\" limit {}".format(range_max, hyperparameter_name, limit_max)
+                self[hyperparameter_name]["range"]["max"] = range_max
+            if nb_values is not None:
+                assert isinstance(nb_values, int) and nb_values >= 2, "Range number of values for hyperparameter \"{}\" must be an integer and >= 2".format(hyperparameter_name)
+
+            if range_min is not None:
+                self[hyperparameter_name]["range"]["min"] = range_min
+            if range_max is not None:
+                self[hyperparameter_name]["range"]["max"] = range_max
+            if nb_values is not None:
+                self[hyperparameter_name]["range"]["nbValues"] = nb_values
+
+        if mode_as_range:
+            self._set_hyperparameter_mode_as_range(hyperparameter_name)
+
+    def _set_categorical_values(self, hyperparameter_name, values=None):
+        self._check_hyperparameter_name(hyperparameter_name)
+        if values is None:
+            warnings.warn("Categorical hyperparameter \"{}\" not modified".format(hyperparameter_name))
+        else:
+            assert isinstance(values, dict), "Invalid values input type for hyperparameter " \
+                                             "\"{}\": ".format(hyperparameter_name) + \
+                                             "must be a dictionary"
+            admissible_values = list(self[hyperparameter_name]["values"].keys())
```

Comment:
Where are these admissible_values first defined ? In the backend ?

Reply:
Yes